File: NandGate.py
from Gate import Gate
from Bus import Bus
from andGate import AND_GATE
from notGate import NotGate

class NandGate(Gate):
    def __init__(self) -> None:
        super().__init__(2, 1)
    
    def process(self):
        b1 = Bus()
        b2 = Bus()
        b3 = Bus()
        b4 = Bus()
        g = AND_GATE()
        n = NotGate()
        g.appendInput(b1)
        g.appendInput(b2)
        g.appendOutput(b3)
        n.appendInput(b3)
        n.appendOutput(b4)

        b1.appendRight(g, 0)
        b2.appendRight(g, 1)
        b3.appendRight(n, 0)


        b1.recieveLeft(self.inputs[0].leftNode)
        b2.recieveLeft(self.inputs[1].leftNode)

        g.process()
        n.process()
        self.values = [ b4.leftNode ]
        self.send(b4.leftNode)

File: javascriptPort/src/library/NandGate.ts
import Bus from "./bus";
import Gate from "./gate";
import AND_GATE from "./AndGate";
import NotGate from "./NotGate";

class NandGate extends Gate {
    values: (number | Gate | null)[] | undefined;
    g: AND_GATE;
    b1: Bus;
    b2: Bus;
    b3: Bus;
    b4: Bus;
    n: NotGate;

    constructor() {
        super(2, 1, "NandGate");
        this.b1 = new Bus();
        this.b2 = new Bus();
        this.b3 = new Bus();
        this.b4 = new Bus();
        this.g = new AND_GATE();
        this.n = new NotGate();

    }
    process() {
        this.g.appendInput(this.b1);
        this.g.appendInput(this.b2);
        this.g.appendOutput(this.b3);
        this.n.appendInput(this.b3);
        this.n.appendOutput(this.b4);
        if(this.inputs[0].leftNode && this.inputs[1].leftNode){
            this.b1.recieveLeft((this.inputs[0].leftNode));
            this.b2.recieveLeft(this.inputs[1].leftNode);
        }
        this.g.process();
        this.n.process();
        this.values = Array.from([ this.b4.leftNode ]);
        this.send( (this.b4.leftNode as number) );
    }
}


export default NandGate